File: app/src/test/java/com/hana053/micropost/presentation/core/components/followbtn/FollowBtnServiceTest.java
package com.hana053.micropost.presentation.core.components.followbtn;

import com.hana053.micropost.interactors.RelationshipInteractor;
import com.hana053.micropost.presentation.core.services.HttpErrorHandler;
import com.hana053.micropost.testing.RobolectricBaseTest;

import org.greenrobot.eventbus.EventBus;
import org.junit.Before;
import org.junit.Test;

import rx.Observable;
import rx.subscriptions.CompositeSubscription;

import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class FollowBtnServiceTest extends RobolectricBaseTest {

    private final RelationshipInteractor relationshipInteractor = mock(RelationshipInteractor.class);

    private FollowBtnService followBtnService;

    @Before
    public void setup() {
        final HttpErrorHandler httpErrorHandler = mock(HttpErrorHandler.class);
        final CompositeSubscription subscriptions = new CompositeSubscription();
        followBtnService = new FollowBtnService(relationshipInteractor, httpErrorHandler, subscriptions);
    }

    @Test
    public void shouldUnfollowWhenClickedFollowBtn() {
        final FollowBtnViewModel viewModel = new FollowBtnViewModel(1, true);
        final FollowBtnClickEvent event = new FollowBtnClickEvent(viewModel);
        when(relationshipInteractor.unfollow(1)).thenReturn(Observable.empty());

        followBtnService.onEvent(event);

        advance();
        verify(relationshipInteractor).unfollow(1);
        assertThat(viewModel.isFollowedByMe.get(), is(false));
    }

    @Test
    public void shouldFollowWhenClickedFollowBtn() {
        final FollowBtnViewModel viewModel = new FollowBtnViewModel(1, false);
        final FollowBtnClickEvent event = new FollowBtnClickEvent(viewModel);
        when(relationshipInteractor.follow(1)).thenReturn(Observable.empty());

        followBtnService.onEvent(event);

        advance();
        verify(relationshipInteractor).follow(1);
        assertThat(viewModel.isFollowedByMe.get(), is(true));
    }

    @Test
    public void shouldSubscribeEventBus() {
        final EventBus eventBus = EventBus.getDefault();
        followBtnService.startObserving();
        assertThat(eventBus.isRegistered(followBtnService), is(true));
        followBtnService.stopObserving();
        assertThat(eventBus.isRegistered(followBtnService), is(false));
    }

}

File: app/src/test/java/com/hana053/micropost/shared/followbtn/FollowBtnServiceTest.kt
package com.hana053.micropost.shared.followbtn

import android.app.Activity
import android.content.Context
import android.widget.Button
import com.github.salomonbrys.kodein.bind
import com.github.salomonbrys.kodein.instance
import com.hana053.micropost.interactors.RelationshipInteractor
import com.hana053.micropost.services.HttpErrorHandler
import com.hana053.micropost.testing.RobolectricBaseTest
import com.hana053.micropost.testing.TestUser
import com.nhaarman.mockito_kotlin.doReturn
import com.nhaarman.mockito_kotlin.mock
import org.hamcrest.CoreMatchers.`is`
import org.hamcrest.MatcherAssert.assertThat
import org.junit.Before
import org.junit.Test
import org.mockito.Mockito.`when`
import org.robolectric.Robolectric
import rx.Observable


class FollowBtnServiceTest : RobolectricBaseTest() {

    private val relationshipInteractor = mock<RelationshipInteractor>()
    private val httpErrorHandler = mock<HttpErrorHandler>()
    private val followBtnService = FollowBtnService(
        relationshipInteractor = relationshipInteractor,
        httpErrorHandler = httpErrorHandler
    )

    @Before
    fun setup() {
        val activity = Robolectric.buildActivity(Activity::class.java).create().get()
        overrideAppBindings {
            bind<Context>() with instance(app)
            bind<Activity>() with instance(activity)
        }
    }

    @Test
    fun shouldFollowOrUnfollow() {
        val userId = 1L
        `when`(relationshipInteractor.follow(userId))
            .doReturn(Observable.just<Void>(null))

        val view = FollowBtnView(Button(app))
        view.render(TestUser.copy(id = userId, isFollowedByMe = false))

        // just confirm that Follow is shown
        assertThat(view.isFollowState(), `is`(true))

        followBtnService.handleFollowBtnClicks(view).subscribe()
        advance()

        assertThat(view.isFollowState(), `is`(false))
    }

    @Test
    fun shouldUnfollow() {
        val userId = 1L
        `when`(relationshipInteractor.unfollow(userId))
            .doReturn(Observable.just<Void>(null))

        val view = FollowBtnView(Button(app))
        view.render(TestUser.copy(id = userId, isFollowedByMe = true))

        // just confirm that Unfollow is shown
        assertThat(view.isFollowState(), `is`(false))

        followBtnService.handleFollowBtnClicks(view).subscribe()
        advance()

        assertThat(view.isFollowState(), `is`(true))
    }
}